# Análise dos dados das variáveis relacionadas ao IDEB

O notebook contém experimentos com essas variáveis. São utilizadas duas bases de dados:<br>
- Váriáveis derivadas do SAEB (informações socioeconômicas)<br> 
- Variáveis com informações das escolas e professores<br>

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
import pandas as pd
from sklearn.feature_selection import SelectKBest, RFE, RFECV, SelectFromModel
from sklearn.feature_selection import mutual_info_classif, f_classif, chi2
from sklearn.preprocessing import LabelEncoder
import numpy as np
from numpy import array 
from sklearn.naive_bayes import MultinomialNB
from sklearn.ensemble import RandomForestClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.svm import SVC
from sklearn.linear_model import LogisticRegression
from sklearn.neighbors import KNeighborsClassifier
from sklearn.preprocessing import MinMaxScaler
from sklearn.model_selection import train_test_split
import sklearn.metrics as metrics

## Importando bases de dados

In [ ]:
#Importando variáveis derivadas do SAEB

#base_estudantes_se = pd.read_csv('C:/Users/maril/Documents/ufs/TCC e Pesquisa IDEB/variaveis_derivadas_se.csv')

#base_estudantes_se = pd.read_csv('/content/drive/MyDrive/TCC/variaveis_derivadas_se.csv')

#base_estudantes_se.set_index('ID_ESCOLA', inplace=True)

#base_estudantes_al = pd.read_csv('C:/Users/maril/Documents/ufs/TCC e Pesquisa IDEB/variaveis_derivadas_al.csv')

base_estudantes_al = pd.read_csv('/content/drive/MyDrive/TCC/variaveis_derivadas_al.csv')

#base_estudantes_al.set_index('ID_ESCOLA', inplace=True)

#base_estudantes_ba = pd.read_csv('/content/drive/MyDrive/TCC/variaveis_derivadas_ba.csv')

dados_pessoais = base_estudantes_al

#base_estudantes_ba.set_index('ID_ESCOLA', inplace=True)

#dados_pessoais = pd.merge(base_estudantes_se, base_estudantes_al, how = 'outer')

#dados_pessoais = pd.merge(base_estudantes_ba, dados_pessoais, how = 'outer')

In [ ]:
dados_pessoais.set_index('ID_ESCOLA', inplace=True)

In [ ]:
dados_pessoais

ID_PROVA_BRASIL  ID_REGIAO  ...  TX_RESP_Q003_J  TX_RESP_Q003_L
ID_ESCOLA                              ...                                
27000095              2017          2  ...               7               7
61188278              2017          2  ...               3               5
27016331              2017          2  ...               7              11
27016366              2017          2  ...               3               4
27016382              2017          2  ...               2               0
...                    ...        ...  ...             ...             ...
61206951              2017          2  ...               1               2
61178197              2017          2  ...               4               3
61197558              2017          2  ...               1               2
61231730              2017          2  ...               2               2
61218351              2017          2  ...              11               8

[244 rows x 235 columns]

**Pré-processamento da base de dados pessoais**

In [ ]:
#Função que verifica se o ID de uma escola é inválido
#Segundo o Dicionário de Dados da base o SAEB 6 - Os códigos dos Municípios e 
#das Escolas que começam com o dígito "6" são máscaras, isto é, são códigos fictícios.
def retira_id_ficticio(id_escola):
    if id_escola[0] == "6":
        return True
    else:
        return False

In [ ]:
#Se retirou as escolas fictícias
for escola in dados_pessoais.index:
    if (retira_id_ficticio(str(escola))):
        dados_pessoais.drop(escola, inplace=True)
        

dados_pessoais

ID_PROVA_BRASIL  ID_REGIAO  ...  TX_RESP_Q003_J  TX_RESP_Q003_L
ID_ESCOLA                              ...                                
27000095              2017          2  ...               7               7
27016331              2017          2  ...               7              11
27016366              2017          2  ...               3               4
27016382              2017          2  ...               2               0
27016412              2017          2  ...               3               5
...                    ...        ...  ...             ...             ...
27024059              2017          2  ...              19              16
27024067              2017          2  ...               6               5
27025047              2017          2  ...              10               6
27025080              2017          2  ...               9              18
27025519              2017          2  ...               6               1

[137 rows x 235 columns]

In [ ]:
#Importando variáveis com informações das escolas e professores
#base_escola_se = pd.read_csv('C:/Users/maril/Documents/ufs/TCC e Pesquisa IDEB/ESCOLAS2017-SE.csv')

##base_escola_se = pd.read_csv('/content/drive/MyDrive/TCC/ESCOLAS2017-SE.csv')

#base_escola_se.set_index('CO_ENTIDADE', inplace=True)

#base_escola_al = pd.read_csv('C:/Users/maril/Documents/ufs/TCC e Pesquisa IDEB/ESCOLAS2017-AL.csv')

base_escola_al = pd.read_csv('/content/drive/MyDrive/TCC/ESCOLAS2017-AL.csv')

base_escola_al.set_index('CO_ENTIDADE', inplace=True)

#base_escola_ba = pd.read_csv('C:/Users/maril/Documents/ufs/TCC e Pesquisa IDEB/ESCOLAS2017-BA.csv')

#base_escola_ba = pd.read_csv('/content/drive/MyDrive/TCC/ESCOLAS2017-BA.csv')

#base_escola_ba.rename(columns={'Unnamed: 0': 'CO_ENTIDADE'}, inplace=True)

#base_escola_ba.set_index('CO_ENTIDADE', inplace=True)

#dados_escola = pd.merge(base_escola_se, base_escola_al, how = 'outer')

#dados_escola = pd.concat([dados_escola, base_escola_ba], join='outer')

dados_escola = base_escola_al

In [ ]:
#dados_escola.set_index('CO_ENTIDADE', inplace=True)

dados_escola

NU_ANO_CENSO  ... Ideb-2017
CO_ENTIDADE                ...          
27000095             2017  ...       3.4
27001237             2017  ...       3.9
27002020             2017  ...       3.3
27002063             2017  ...       3.8
27002950             2017  ...       3.5
...                   ...  ...       ...
27226697             2017  ...       4.3
27226808             2017  ...       3.9
27227189             2017  ...       5.7
27235009             2017  ...       5.1
27353036             2017  ...       3.0

[143 rows x 184 columns]

In [ ]:
col_ideb = dados_escola['Ideb-2017']

dados_escola.drop('Ideb-2017', axis=1, inplace=True)

In [ ]:
col_ideb

CO_ENTIDADE
27000095    3.4
27001237    3.9
27002020    3.3
27002063    3.8
27002950    3.5
           ... 
27226697    4.3
27226808    3.9
27227189    5.7
27235009    5.1
27353036    3.0
Name: Ideb-2017, Length: 143, dtype: float64

In [ ]:
dados_escola

NU_ANO_CENSO  ... QtdeAlunosTotal
CO_ENTIDADE                ...                
27000095             2017  ...             508
27001237             2017  ...             792
27002020             2017  ...             652
27002063             2017  ...             309
27002950             2017  ...             442
...                   ...  ...             ...
27226697             2017  ...             480
27226808             2017  ...             397
27227189             2017  ...             985
27235009             2017  ...             627
27353036             2017  ...             865

[143 rows x 183 columns]

#### Pré-processamento da base das escolas

In [ ]:
#Retirando valores ausentes
filteredColumns = dados_escola.dtypes[dados_escola.dtypes == np.float64]

listOfColumnNames = list(filteredColumns.index)

# retorna o número de valores ausentes para cada coluna
dados_escola[listOfColumnNames].isnull().sum()

CO_ORGAO_REGIONAL                3
TP_CATEGORIA_ESCOLA_PRIVADA    118
IN_CONVENIADA_PP               118
TP_CONVENIO_PODER_PUBLICO      138
IN_MANT_ESCOLA_PRIVADA_EMP     118
                              ... 
asQtdeDocentesEfetivos          29
QtdeDocentesTempo               25
QtdeDocentesTerceir            135
QtdedocentesCLT                143
QtdeDocentesComLicenciatura      0
Length: 160, dtype: int64

In [ ]:
#Retirando colunas que apresentam mais de 60% do valores nulos
dados_escola.dropna(thresh=len(dados_escola)*0.6, axis=1, inplace=True)

#Preenchendo valores ausentes
dados_escola.fillna(0, inplace=True)

filteredColumns = dados_escola.dtypes[dados_escola.dtypes == np.float64]

listOfColumnNames = list(filteredColumns.index)

# retorna o número de valores ausentes para cada coluna
dados_escola[listOfColumnNames].isnull().sum()

CO_ORGAO_REGIONAL               0
TP_REGULAMENTACAO               0
IN_LOCAL_FUNC_PREDIO_ESCOLAR    0
TP_OCUPACAO_PREDIO_ESCOLAR      0
IN_LOCAL_FUNC_SALAS_EMPRESA     0
                               ..
IN_ESP_EXCLUSIVA_PROF           0
QtdeDocentes                    0
asQtdeDocentesEfetivos          0
QtdeDocentesTempo               0
QtdeDocentesComLicenciatura     0
Length: 143, dtype: int64

In [ ]:
dados_escola[listOfColumnNames] = dados_escola[listOfColumnNames].astype(int)

In [ ]:
dados_escola.insert(len(dados_escola.columns), 'Ideb-2017', col_ideb)
#['Ideb-2017'] = col_ideb

/usr/local/lib/python3.7/dist-packages/IPython/core/interactiveshell.py:2828: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead.  To get a de-fragmented frame, use `newframe = frame.copy()`
  if self.run_code(code, result):


In [ ]:
dados_escola

NU_ANO_CENSO  ... Ideb-2017
CO_ENTIDADE                ...          
27000095             2017  ...       3.4
27001237             2017  ...       3.9
27002020             2017  ...       3.3
27002063             2017  ...       3.8
27002950             2017  ...       3.5
...                   ...  ...       ...
27226697             2017  ...       4.3
27226808             2017  ...       3.9
27227189             2017  ...       5.7
27235009             2017  ...       5.1
27353036             2017  ...       3.0

[143 rows x 167 columns]

In [ ]:
#Colunas categóricas
filteredColumns = dados_escola.dtypes[dados_escola.dtypes == np.object]

listOfColumnNames = list(filteredColumns.index)

dados_escola[listOfColumnNames]

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:2: DeprecationWarning: `np.object` is a deprecated alias for the builtin `object`. To silence this warning, use `object` by itself. Doing this will not modify any behavior and is safe. 
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
  


NO_ENTIDADE  ... DT_ANO_LETIVO_TERMINO
CO_ENTIDADE                                                     ...                      
27000095     ESCOLA ESTADUAL MONSENHOR SEBASTIAO ALVES BEZERRA  ...            15/12/2017
27001237                           ESCOLA ESTADUAL LUIZ BASTOS  ...            15/12/2017
27002020           ESCOLA ESTADUAL GENTIL DE ALBUQUERQUE MALTA  ...            15/12/2017
27002063                 ESCOLA ESTADUAL SANTA CRUZ DO DESERTO  ...            15/12/2017
27002950       ESCOLA ESTADUAL DE EDUCACAO BASICA DE PARICONHA  ...            15/12/2017
...                                                        ...  ...                   ...
27226697                                 COLEGIO ROSALVO FELIX  ...            22/12/2017
27226808          ESCOLA ESTADUAL PROFA ELZA SOARES CAVALCANTE  ...            14/12/2017
27227189            COLEGIO DE EDUCACAO BASICA SANTA ESMERALDA  ...            16/12/2017
27235009                 COLEGIO DE EDUCACAO BASICA SANTA AFRA  ...            15/12/2017
27353036     ESCOLA ESTADUAL DE ED BASICA E PROFISSIONAL JO...  ...            29/12/2017

[143 rows x 3 columns]

In [ ]:
#Tratando datas, convertendo para timestamp

inicio = pd.to_datetime(dados_escola['DT_ANO_LETIVO_INICIO'], infer_datetime_format = True)
termino = pd.to_datetime(dados_escola['DT_ANO_LETIVO_TERMINO'], infer_datetime_format = True)

dados_escola['DT_ANO_LETIVO_INICIO'] = inicio.values.astype (np.int64)
dados_escola['DT_ANO_LETIVO_TERMINO'] = termino.values.astype (np.int64)

In [ ]:
dados_escola.drop('NO_ENTIDADE', axis=1, inplace=True)

In [ ]:
dados_escola.columns.to_list()

['NU_ANO_CENSO',
 'CO_ORGAO_REGIONAL',
 'TP_SITUACAO_FUNCIONAMENTO',
 'DT_ANO_LETIVO_INICIO',
 'DT_ANO_LETIVO_TERMINO',
 'CO_REGIAO',
 'CO_MESORREGIAO',
 'CO_MICRORREGIAO',
 'CO_UF',
 'CO_MUNICIPIO',
 'CO_DISTRITO',
 'TP_DEPENDENCIA',
 'TP_LOCALIZACAO',
 'TP_REGULAMENTACAO',
 'IN_LOCAL_FUNC_PREDIO_ESCOLAR',
 'TP_OCUPACAO_PREDIO_ESCOLAR',
 'IN_LOCAL_FUNC_SALAS_EMPRESA',
 'IN_LOCAL_FUNC_SOCIOEDUCATIVO',
 'IN_LOCAL_FUNC_UNID_PRISIONAL',
 'IN_LOCAL_FUNC_PRISIONAL_SOCIO',
 'IN_LOCAL_FUNC_TEMPLO_IGREJA',
 'IN_LOCAL_FUNC_CASA_PROFESSOR',
 'IN_LOCAL_FUNC_GALPAO',
 'IN_LOCAL_FUNC_SALAS_OUTRA_ESC',
 'IN_LOCAL_FUNC_OUTROS',
 'IN_PREDIO_COMPARTILHADO',
 'IN_AGUA_FILTRADA',
 'IN_AGUA_REDE_PUBLICA',
 'IN_AGUA_POCO_ARTESIANO',
 'IN_AGUA_CACIMBA',
 'IN_AGUA_FONTE_RIO',
 'IN_AGUA_INEXISTENTE',
 'IN_ENERGIA_REDE_PUBLICA',
 'IN_ENERGIA_GERADOR',
 'IN_ENERGIA_OUTROS',
 'IN_ENERGIA_INEXISTENTE',
 'IN_ESGOTO_REDE_PUBLICA',
 'IN_ESGOTO_FOSSA',
 'IN_ESGOTO_INEXISTENTE',
 'IN_LIXO_COLETA_PERIODICA',
 'IN_LIXO_

In [ ]:
dados_pessoais

ID_PROVA_BRASIL  ID_REGIAO  ...  TX_RESP_Q003_J  TX_RESP_Q003_L
ID_ESCOLA                              ...                                
27000095              2017          2  ...               7               7
27016331              2017          2  ...               7              11
27016366              2017          2  ...               3               4
27016382              2017          2  ...               2               0
27016412              2017          2  ...               3               5
...                    ...        ...  ...             ...             ...
27024059              2017          2  ...              19              16
27024067              2017          2  ...               6               5
27025047              2017          2  ...              10               6
27025080              2017          2  ...               9              18
27025519              2017          2  ...               6               1

[137 rows x 235 columns]

### Unindo as bases de dados 

Foi realizado join nos dois datasets, nesse momento se percebeu que não existem variáveis escolas particulares na base derivada do SAEB

In [ ]:
dataset = dados_pessoais.join(dados_escola, how="inner")

In [ ]:
dataset

ID_PROVA_BRASIL  ID_REGIAO  ...  QtdeAlunosTotal  Ideb-2017
27000095             2017          2  ...              508        3.4
27001237             2017          2  ...              792        3.9
27002020             2017          2  ...              652        3.3
27002063             2017          2  ...              309        3.8
27002950             2017          2  ...              442        3.5
...                   ...        ...  ...              ...        ...
27226131             2017          2  ...              903        2.8
27226239             2017          2  ...              976        3.1
27226468             2017          2  ...              597        4.1
27226808             2017          2  ...              397        3.9
27353036             2017          2  ...              865        3.0

[118 rows x 401 columns]

In [ ]:
dados_escola.dtypes[dados_escola.dtypes == np.float64]

Ideb-2017    float64
dtype: object

In [ ]:
dataset['ID_DEPENDENCIA_ADM'].unique()

array([2, 1])

In [ ]:
#dataset.to_csv('C:/Users/maril/Documents/ufs/TCC e Pesquisa Ideb/variaveis_derivadas_escola_estudante_5.0.csv')
dataset.to_csv('/content/drive/MyDrive/TCC/variaveis_derivadas_escola_estudante_5.0.csv')

### Discretização dos rótulos, nota do Ideb

**Discretização:** É o processo de transferência de funções contínuas, modelos, variáveis e equações em contrapartes discretas.

In [ ]:
# São três classes média, acima da média e abaixo da média
def discretizacao(ideb, mean_ideb):
    classe = []
    for nota in ideb['Ideb-2017']:
        if (mean_ideb == nota):
            classe.append('média')
        elif (mean_ideb < nota):
            classe.append('acima da média')
        else:
            classe.append('abaixo da média')
    return classe

In [ ]:
#discretizando as notas do Ideb

Ideb = dataset[['Ideb-2017']]

mean_Ideb = round(Ideb.mean(),1)

classe_ideb = discretizacao(Ideb, mean_Ideb['Ideb-2017'])

Ideb.insert(1, 'classe-Ideb', classe_ideb)

Ideb

Ideb-2017      classe-Ideb
27000095        3.4            média
27001237        3.9   acima da média
27002020        3.3  abaixo da média
27002063        3.8   acima da média
27002950        3.5   acima da média
...             ...              ...
27226131        2.8  abaixo da média
27226239        3.1  abaixo da média
27226468        4.1   acima da média
27226808        3.9   acima da média
27353036        3.0  abaixo da média

[118 rows x 2 columns]

In [ ]:
#Variavel a ser prevista
y = Ideb['classe-Ideb']

**Seleção de Atributos:**  É um passo da fase de pré-processamento que tem a finalidade de selecionar o melhor subconjunto de atributos que pode substituir o conjunto de dados original. 

https://www.analyticsvidhya.com/blog/2016/12/introduction-to-feature-selection-methods-with-an-example-or-how-to-select-the-right-variables/#:~:text=Embedded%20methods%20combine%20the%20qualities,penalization%20functions%20to%20reduce%20overfitting.

Abordagens:

- Filtro: a relevância dos atributos é avaliada independentemente – não
envolve o uso de um método de AM ;

- Wrapper (Envólucro): Avalia o subconjunto de atributos arti culado ao método de AM que realizará a indução do conceito;

- Embedded (Embutidos): Os métodos incorporados combinam as qualidades dos métodos de filtro e wrapper. Ele é implementado por algoritmos que possuem seus próprios métodos de seleção de recursos integrados.

In [ ]:
#Filter based
# f_classif: ANOVA F-value between label/feature for classification tasks.
#ANOVA stands for Analysis of variance. It is similar to LDA except for the fact that it is operated using one or more categorical 
#independent features and one continuous dependent feature. It provides a statistical test 
#of whether the means of several groups are equal or not.
def selecao_atributos(X, y, quant):
    select = SelectKBest(score_func=f_classif, k=quant)
    #select = SelectKBest(score_func=chi2, k=quant)
    z = select.fit_transform(X, y)
    
    filter = select.get_support(indices=True)
    features = array(X.columns)
    
    return X.iloc[:, filter]

In [ ]:
#Wrapper-based
#from sklearn.cluster import KMeans

def wrapper_based(X, X_transform, y, quant):

  #select_features = RFE(estimator=KMeans(n_clusters=3), n_features_to_select=100)
  select_features = RFE(estimator=SVC(kernel="linear"), n_features_to_select=100)
  '''select_features = RFECV(
      estimator=SVC(kernel="linear"),
      min_features_to_select= quant,
      cv = 5
  )'''
  select_features.fit(X_transform, y)

  filter = select_features.get_support(indices=True)
  return X.iloc[:, filter]

In [ ]:
# Embedded-based

def embedded_based(X, X_transform, y, quant):

  select_features = SelectFromModel(LogisticRegression(), max_features=quant)
  #select_features = SelectFromModel(DecisionTreeClassifier(), max_features=quant)
  #select_features = SelectFromModel(SVC(kernel="rbf"), max_features=quant)
  #select_features = SelectFromModel(RandomForestClassifier(), max_features=quant)
  select_features.fit(X_transform, y)

  filter = select_features.get_support(indices=True)

  return X.iloc[:, filter]

**Métricas de Avaliação**


In [ ]:
def acuracia(resultados, classes):
    #resultados = cross_val_predict(modelo, textos, classes, cv=10)  
    return 'Acurácia do modelo: {} %'.format(round(metrics.accuracy_score(classes, resultados)*100, 2))

**Normalização:** é mudar os valores das colunas numéricas no conjunto de dados para usar uma escala comum, sem distorcer as diferenças nos intervalos de valores nem perder informações.

- **Normalização MinMax:** Para cada recurso, o valor mínimo desse recurso é transformado em 0, o valor máximo é transformado em 1 e todos os outros valores são transformados em um decimal entre 0 e 1.

In [ ]:
def minmax(data):
    scaler = MinMaxScaler()
    data_scaler = scaler.fit_transform(data)
    return data_scaler

## Criação dos modelos

**Serão feitas três análises:** Apenas com as variáveis socioeconômicas, com as variáveis de escolas e professores e depois com o dataset completo

Para cada uma serão criados modelos com todos os atributos e também com seleção de atributos, a fim de verificar o de maior acurácia

In [ ]:
def inicializar_modelos():
    naive = MultinomialNB()
    random = RandomForestClassifier(n_estimators=150)
    decision_tree = DecisionTreeClassifier()
    svm = SVC(kernel="linear")
    r_logistic = LogisticRegression(max_iter=250)
    knn = KNeighborsClassifier(n_neighbors=3)
    return naive, random, decision_tree, svm, r_logistic, knn

#### -Primeira análise: variáveis socioeconômicas derivadas da prova SAEB**

In [ ]:
#Foi retirada colunas do 'ID_PROVA_BRASIL', 'ID_REGIAO', 'ID_UF'
columns = dataset.columns[3:234]

columns

Index(['ID_MUNICIPIO', 'ID_AREA', 'ID_DEPENDENCIA_ADM', 'ID_LOCALIZACAO',
       'IN_SITUACAO_CENSO', 'TX_RESP_Q001_A', 'TX_RESP_Q001_B',
       'TX_RESP_Q020_A', 'TX_RESP_Q020_B', 'TX_RESP_Q021_A',
       ...
       'TX_RESP_Q003_A', 'TX_RESP_Q003_B', 'TX_RESP_Q003_C', 'TX_RESP_Q003_D',
       'TX_RESP_Q003_E', 'TX_RESP_Q003_F', 'TX_RESP_Q003_G', 'TX_RESP_Q003_H',
       'TX_RESP_Q003_I', 'TX_RESP_Q003_J'],
      dtype='object', length=231)

In [ ]:
X =  dataset[columns]

In [ ]:
#Questões de sexo, raça, ano, mes de nascimento, conformidade com censo
question = ['TX_RESP_Q001_A', 'TX_RESP_Q001_B', 'TX_RESP_Q002_A', 'TX_RESP_Q002_B', 'TX_RESP_Q002_C', 'TX_RESP_Q002_D', 
            'TX_RESP_Q002_E', 'TX_RESP_Q002_F', 'TX_RESP_Q003_A', 'TX_RESP_Q003_B', 'TX_RESP_Q003_C', 'TX_RESP_Q003_D', 
            'TX_RESP_Q003_E', 'TX_RESP_Q003_F', 'TX_RESP_Q003_G', 'TX_RESP_Q003_H', 'TX_RESP_Q003_I', 'TX_RESP_Q003_J', 
            'TX_RESP_Q004_A', 'TX_RESP_Q004_B', 'TX_RESP_Q004_C', 'TX_RESP_Q004_D', 
            'TX_RESP_Q004_E', 'TX_RESP_Q004_F', 'TX_RESP_Q004_G', 'TX_RESP_Q004_H', 'IN_SITUACAO_CENSO']

delete_at = ['IN_SITUACAO_CENSO']

X.drop(delete_at, axis=1, inplace=True)

/usr/local/lib/python3.7/dist-packages/pandas/core/frame.py:4913: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  errors=errors,


In [ ]:
filteredColumns = X.dtypes[X.dtypes == np.float]

listOfColumnNames = list(filteredColumns.index)

X[listOfColumnNames]

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: DeprecationWarning: `np.float` is a deprecated alias for the builtin `float`. To silence this warning, use `float` by itself. Doing this will not modify any behavior and is safe. If you specifically wanted the numpy scalar type, use `np.float64` here.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
  """Entry point for launching an IPython kernel.


Empty DataFrame
Columns: []
Index: [27000095, 27001237, 27002020, 27002063, 27002950, 27003108, 27003221, 27003515, 27003973, 27004490, 27005224, 27007120, 27007456, 27008827, 27009998, 27010074, 27010090, 27010767, 27011100, 27011844, 27012115, 27012166, 27012611, 27012620, 27012638, 27013405, 27013871, 27013979, 27014002, 27014010, 27014029, 27014037, 27014932, 27014959, 27015068, 27015114, 27015920, 27016331, 27016366, 27016382, 27016412, 27016420, 27016471, 27017702, 27018016, 27018016, 27018350, 27018962, 27018997, 27019608, 27019608, 27020274, 27020320, 27020819, 27021467, 27022277, 27022838, 27023516, 27023583, 27024016, 27024059, 27024067, 27025047, 27025080, 27025519, 27026337, 27026370, 27026744, 27027082, 27027597, 27028380, 27029123, 27029166, 27029336, 27030245, 27030806, 27031063, 27031365, 27032280, 27033384, 27034070, 27034240, 27034887, 27035786, 27035948, 27035980, 27036499, 27036642, 27036677, 27037037, 27037088, 27038297, 27038556, 27038637, 27038734, 27040240, 27040348, 27040860, 27040984, 27043436, ...]

[118 rows x 0 columns]

**Divisão entre conjunto de treino e teste**

In [ ]:
X_scaler = minmax(X)

X_train, X_test, y_train, y_test = train_test_split(X_scaler, y, random_state=50, test_size=0.2)

**Fase de Treinamento e Teste**

In [ ]:
naive, random, decision_tree, svm, r_logistic, knn = inicializar_modelos()

In [ ]:
naive.fit(X_train, y_train)
random.fit(X_train, y_train)
decision_tree.fit(X_train, y_train)
svm.fit(X_train, y_train)
r_logistic.fit(X_train, y_train)
knn.fit(X_train, y_train)

KNeighborsClassifier(n_neighbors=3)

In [ ]:
random_result = random.predict(X_test)
naive_result = naive.predict(X_test)
decision_result = decision_tree.predict(X_test)
svm_result = svm.predict(X_test)
logistic_result = r_logistic.predict(X_test)
knn_result = knn.predict(X_test)

print("Random forest: " + acuracia(random_result, y_test))
print("Naive Bayes: " + acuracia(naive_result, y_test))
print("Decision Tree: " + acuracia(decision_result, y_test))
print("SVM: " + acuracia(svm_result, y_test))
print("Logistic Regression: "+acuracia(logistic_result, y_test))
print("KNN: "+acuracia(knn_result, y_test))

Random forest: Acurácia do modelo: 58.33 %
Naive Bayes: Acurácia do modelo: 50.0 %
Decision Tree: Acurácia do modelo: 45.83 %
SVM: Acurácia do modelo: 41.67 %
Logistic Regression: Acurácia do modelo: 41.67 %
KNN: Acurácia do modelo: 50.0 %


**-Primeira Análise: Aplicando Seleção de Atributos (Filtragem)**

In [ ]:
X_selection = selecao_atributos(X, y, 100)

X_selection

ID_DEPENDENCIA_ADM  TX_RESP_Q001_B  ...  TX_RESP_Q003_D  TX_RESP_Q003_F
27000095                   2              59  ...               7              11
27001237                   2             108  ...              14              12
27002020                   2              60  ...               8               6
27002063                   2              24  ...               2               4
27002950                   2              72  ...              11              16
...                      ...             ...  ...             ...             ...
27226131                   2              86  ...              14               9
27226239                   2              67  ...               6              11
27226468                   2              90  ...               9              14
27226808                   2              46  ...               3               7
27353036                   2              86  ...              14              17

[118 rows x 100 columns]

In [ ]:
X_scaler = minmax(X_selection)

X_train, X_test, y_train, y_test = train_test_split(X_scaler, y, random_state=50, test_size=0.2)

**Fase de Treinamento e Teste**

In [ ]:
naive, random, decision_tree, svm, r_logistic, knn = inicializar_modelos()

In [ ]:
naive.fit(X_train, y_train)
random.fit(X_train, y_train)
decision_tree.fit(X_train, y_train)
svm.fit(X_train, y_train)
r_logistic.fit(X_train, y_train)
knn.fit(X_train, y_train)

KNeighborsClassifier(n_neighbors=3)

In [ ]:
random_result = random.predict(X_test)
naive_result = naive.predict(X_test)
decision_result = decision_tree.predict(X_test)
svm_result = svm.predict(X_test)
logistic_result = r_logistic.predict(X_test)
knn_result = knn.predict(X_test)

print("Random forest: " + acuracia(random_result, y_test))
print("Naive Bayes: " + acuracia(naive_result, y_test))
print("Decision Tree: " + acuracia(decision_result, y_test))
print("SVM: " + acuracia(svm_result, y_test))
print("Logistic Regression: "+acuracia(logistic_result, y_test))
print("KNN: "+acuracia(knn_result, y_test))

Random forest: Acurácia do modelo: 45.83 %
Naive Bayes: Acurácia do modelo: 54.17 %
Decision Tree: Acurácia do modelo: 41.67 %
SVM: Acurácia do modelo: 45.83 %
Logistic Regression: Acurácia do modelo: 58.33 %
KNN: Acurácia do modelo: 33.33 %


**-Primeira Análise: Aplicando Seleção de Atributos (Embutida)**

In [ ]:
X_embedded = embedded_based(X, X_scaler, y, 100)

X_embedded

ID_MUNICIPIO  ID_DEPENDENCIA_ADM  ...  TX_RESP_Q039_C  TX_RESP_Q039_D
27000095       2700102                   2  ...              32               2
27001237       2701605                   2  ...              53               1
27002020       2705002                   2  ...              28               3
27002063       2705002                   2  ...              15               2
27002950       2706422                   2  ...              15               7
...                ...                 ...  ...             ...             ...
27226131       2700706                   2  ...              24               0
27226239       2702306                   2  ...              24               0
27226468       2708303                   2  ...              55               3
27226808       2707602                   2  ...              17               3
27353036       2709152                   2  ...              25               3

[118 rows x 36 columns]

In [ ]:
X_scaler_em = minmax(X_embedded)

X_train, X_test, y_train, y_test = train_test_split(X_scaler_em, y, random_state=50, test_size=0.2)

In [ ]:
naive, random, decision_tree, svm, r_logistic, knn = inicializar_modelos()

In [ ]:
naive.fit(X_train, y_train)
random.fit(X_train, y_train)
decision_tree.fit(X_train, y_train)
svm.fit(X_train, y_train)
r_logistic.fit(X_train, y_train)
knn.fit(X_train, y_train)

KNeighborsClassifier(n_neighbors=3)

In [ ]:
random_result = random.predict(X_test)
naive_result = naive.predict(X_test)
decision_result = decision_tree.predict(X_test)
svm_result = svm.predict(X_test)
logistic_result = r_logistic.predict(X_test)
knn_result = knn.predict(X_test)

print("Random forest: " + acuracia(random_result, y_test))
print("Naive Bayes: " + acuracia(naive_result, y_test))
print("Decision Tree: " + acuracia(decision_result, y_test))
print("SVM: " + acuracia(svm_result, y_test))
print("Logistic Regression: "+acuracia(logistic_result, y_test))
print("KNN: "+acuracia(knn_result, y_test))

Random forest: Acurácia do modelo: 62.5 %
Naive Bayes: Acurácia do modelo: 54.17 %
Decision Tree: Acurácia do modelo: 45.83 %
SVM: Acurácia do modelo: 62.5 %
Logistic Regression: Acurácia do modelo: 62.5 %
KNN: Acurácia do modelo: 54.17 %


**-Primeira Análise: Aplicando Seleção de Atributos (Wrapper)**

In [ ]:
X_wrapper = wrapper_based(X, X_scaler, y, 100)

X_wrapper

ID_MUNICIPIO  ID_AREA  ...  TX_RESP_Q045_C  TX_RESP_Q045_D
27000095       2700102        2  ...               1               1
27001237       2701605        2  ...               7               9
27002020       2705002        2  ...               3               2
27002063       2705002        2  ...               2               1
27002950       2706422        2  ...               1               4
...                ...      ...  ...             ...             ...
27226131       2700706        2  ...               3               1
27226239       2702306        2  ...               2               4
27226468       2708303        2  ...               5               7
27226808       2707602        2  ...               3               0
27353036       2709152        2  ...               1              11

[118 rows x 100 columns]

In [ ]:
X_scaler_w = minmax(X_wrapper)

X_train, X_test, y_train, y_test = train_test_split(X_scaler_w, y, random_state=50, test_size=0.2)

In [ ]:
naive, random, decision_tree, svm, r_logistic, knn = inicializar_modelos()

In [ ]:
naive.fit(X_train, y_train)
random.fit(X_train, y_train)
decision_tree.fit(X_train, y_train)
svm.fit(X_train, y_train)
r_logistic.fit(X_train, y_train)
knn.fit(X_train, y_train)

KNeighborsClassifier(n_neighbors=3)

In [ ]:
random_result = random.predict(X_test)
naive_result = naive.predict(X_test)
decision_result = decision_tree.predict(X_test)
svm_result = svm.predict(X_test)
logistic_result = r_logistic.predict(X_test)
knn_result = knn.predict(X_test)

print("Random forest: " + acuracia(random_result, y_test))
print("Naive Bayes: " + acuracia(naive_result, y_test))
print("Decision Tree: " + acuracia(decision_result, y_test))
print("SVM: " + acuracia(svm_result, y_test))
print("Logistic Regression: "+acuracia(logistic_result, y_test))
print("KNN: "+acuracia(knn_result, y_test))

Random forest: Acurácia do modelo: 50.0 %
Naive Bayes: Acurácia do modelo: 62.5 %
Decision Tree: Acurácia do modelo: 41.67 %
SVM: Acurácia do modelo: 58.33 %
Logistic Regression: Acurácia do modelo: 58.33 %
KNN: Acurácia do modelo: 41.67 %


#### -Segunda análise: Análise com o dataset completo

In [ ]:
#Lembrar de tratar datas

'''question = ['TX_RESP_Q001_A', 'TX_RESP_Q001_B', 'TX_RESP_Q002_A', 'TX_RESP_Q002_B', 'TX_RESP_Q002_C', 'TX_RESP_Q002_D', 
            'TX_RESP_Q002_E', 'TX_RESP_Q002_F', 'TX_RESP_Q003_A', 'TX_RESP_Q003_B', 'TX_RESP_Q003_C', 'TX_RESP_Q003_D', 
            'TX_RESP_Q003_E', 'TX_RESP_Q003_F', 'TX_RESP_Q003_G', 'TX_RESP_Q003_H', 'TX_RESP_Q003_I', 'TX_RESP_Q003_J', 
            'TX_RESP_Q003_L', 'TX_RESP_Q004_A', 'TX_RESP_Q004_B', 'TX_RESP_Q004_C', 'TX_RESP_Q004_D', 'CO_REGIAO',
            'TP_SITUACAO_FUNCIONAMENTO', 'TP_REGULAMENTACAO',
            'CO_MESORREGIAO', 'CO_MICRORREGIAO', 'CO_DISTRITO', 'TP_LOCALIZACAO',  'CO_ORGAO_REGIONAL', 'TP_DEPENDENCIA',
            'TX_RESP_Q004_E', 'TX_RESP_Q004_F', 'TX_RESP_Q004_G', 'TX_RESP_Q004_H', 'ID_PROVA_BRASIL', 'ID_REGIAO', 'ID_UF', 'ID_MUNICIPIO', 
            'DT_ANO_LETIVO_INICIO', 'DT_ANO_LETIVO_TERMINO', 'IN_SITUACAO_CENSO', 'NU_ANO_CENSO', 'CO_UF', 'CO_MUNICIPIO', 'Ideb-2017']'''

'''delete = ['CO_MESORREGIAO', 'CO_MICRORREGIAO', 'CO_DISTRITO', 'TP_LOCALIZACAO',  'CO_ORGAO_REGIONAL', 'TP_DEPENDENCIA',
          'ID_PROVA_BRASIL', 'ID_REGIAO', 'ID_UF', 'ID_MUNICIPIO', 'IN_SITUACAO_CENSO', 'Ideb-2017']'''


#X = dataset.drop(delete, axis=1)

'''X = dataset.drop(['Ideb-2017','IN_SITUACAO_CENSO', 'TX_RESP_Q001_A', 'TX_RESP_Q001_B', 'TX_RESP_Q002_A', 'TX_RESP_Q002_B', 'TX_RESP_Q002_C', 'TX_RESP_Q002_D', 
            'TX_RESP_Q002_E', 'TX_RESP_Q002_F', 'TX_RESP_Q003_A', 'TX_RESP_Q003_B', 'TX_RESP_Q003_C', 'TX_RESP_Q003_D', 
            'TX_RESP_Q003_E', 'TX_RESP_Q003_F', 'TX_RESP_Q003_G', 'TX_RESP_Q003_H', 'TX_RESP_Q003_I', 'TX_RESP_Q003_J', 
            'TX_RESP_Q003_L', 'TX_RESP_Q004_A', 'TX_RESP_Q004_B', 'TX_RESP_Q004_C', 'TX_RESP_Q004_D',], axis=1)'''

X = dataset.drop(['Ideb-2017','IN_SITUACAO_CENSO'], axis=1)

In [ ]:
list(X.columns)

['ID_PROVA_BRASIL',
 'ID_REGIAO',
 'ID_UF',
 'ID_MUNICIPIO',
 'ID_AREA',
 'ID_DEPENDENCIA_ADM',
 'ID_LOCALIZACAO',
 'TX_RESP_Q001_A',
 'TX_RESP_Q001_B',
 'TX_RESP_Q020_A',
 'TX_RESP_Q020_B',
 'TX_RESP_Q021_A',
 'TX_RESP_Q021_B',
 'TX_RESP_Q024_A',
 'TX_RESP_Q024_B',
 'TX_RESP_Q025_A',
 'TX_RESP_Q025_B',
 'TX_RESP_Q027_A',
 'TX_RESP_Q027_B',
 'TX_RESP_Q028_A',
 'TX_RESP_Q028_B',
 'TX_RESP_Q029_A',
 'TX_RESP_Q029_B',
 'TX_RESP_Q030_A',
 'TX_RESP_Q030_B',
 'TX_RESP_Q031_A',
 'TX_RESP_Q031_B',
 'TX_RESP_Q038_A',
 'TX_RESP_Q038_B',
 'TX_RESP_Q043_A',
 'TX_RESP_Q043_B',
 'TX_RESP_Q044_A',
 'TX_RESP_Q044_B',
 'TX_RESP_Q052_A',
 'TX_RESP_Q052_B',
 'TX_RESP_Q018_A',
 'TX_RESP_Q018_B',
 'TX_RESP_Q018_C',
 'TX_RESP_Q022_A',
 'TX_RESP_Q022_B',
 'TX_RESP_Q022_C',
 'TX_RESP_Q026_A',
 'TX_RESP_Q026_B',
 'TX_RESP_Q026_C',
 'TX_RESP_Q032_A',
 'TX_RESP_Q032_B',
 'TX_RESP_Q032_C',
 'TX_RESP_Q033_A',
 'TX_RESP_Q033_B',
 'TX_RESP_Q033_C',
 'TX_RESP_Q034_A',
 'TX_RESP_Q034_B',
 'TX_RESP_Q034_C',
 'TX_RESP_Q

In [ ]:
X

ID_PROVA_BRASIL  ...  QtdeAlunosTotal
27000095             2017  ...              508
27001237             2017  ...              792
27002020             2017  ...              652
27002063             2017  ...              309
27002950             2017  ...              442
...                   ...  ...              ...
27226131             2017  ...              903
27226239             2017  ...              976
27226468             2017  ...              597
27226808             2017  ...              397
27353036             2017  ...              865

[118 rows x 399 columns]

In [ ]:
filteredColumns = X.dtypes[X.dtypes == np.float64]

listOfColumnNames = list(filteredColumns.index)

#X[listOfColumnNames].fillna(-1, inplace=True)

#X[listOfColumnNames] = X[listOfColumnNames].astype(int)

X[listOfColumnNames].isnull().sum()

Series([], dtype: float64)

In [ ]:
X_transform = minmax(X)

X_train, X_test, y_train, y_test = train_test_split(X_transform, y, random_state=50, test_size=0.2)

**Fase de Treinamento e Teste**

In [ ]:
naive, random, decision_tree, svm, r_logistic, knn = inicializar_modelos()

In [ ]:
naive.fit(X_train, y_train)
random.fit(X_train, y_train)
decision_tree.fit(X_train, y_train)
svm.fit(X_train, y_train)
r_logistic.fit(X_train, y_train)
knn.fit(X_train, y_train)

KNeighborsClassifier(n_neighbors=3)

In [ ]:
random_result = random.predict(X_test)
naive_result = naive.predict(X_test)
decision_result = decision_tree.predict(X_test)
svm_result = svm.predict(X_test)
logistic_result = r_logistic.predict(X_test)
knn_result = knn.predict(X_test)

print("Random forest: " + acuracia(random_result, y_test))
print("Naive Bayes: " + acuracia(naive_result, y_test))
print("Decision Tree: " + acuracia(decision_result, y_test))
print("SVM: " + acuracia(svm_result, y_test))
print("Logistic Regression: "+acuracia(logistic_result, y_test))
print("KNN: "+acuracia(knn_result, y_test))

Random forest: Acurácia do modelo: 50.0 %
Naive Bayes: Acurácia do modelo: 62.5 %
Decision Tree: Acurácia do modelo: 54.17 %
SVM: Acurácia do modelo: 41.67 %
Logistic Regression: Acurácia do modelo: 54.17 %
KNN: Acurácia do modelo: 50.0 %


**-Segunda Análise: Aplicando Seleção de Atributos (Embutida)**

In [ ]:
X_embedded = embedded_based(X, X_transform, y, 100)

X_embedded

/usr/local/lib/python3.7/dist-packages/sklearn/linear_model/_logistic.py:818: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG,


TX_RESP_Q030_B  TX_RESP_Q026_C  ...  IN_COMUM_EJA_MEDIO  QtdeDocentesFCEM
27000095               1               9  ...                   1                 0
27001237               4              19  ...                   1                31
27002020               5               9  ...                   1                 0
27002063               1               9  ...                   0                 0
27002950               4              14  ...                   1                 0
...                  ...             ...  ...                 ...               ...
27226131               4              30  ...                   1                 5
27226239               2              26  ...                   1                 0
27226468               4              13  ...                   1                 0
27226808               1               7  ...                   1                 6
27353036               4              17  ...                   0                59

[118 rows x 100 columns]

In [ ]:
X_transform_em = minmax(X_embedded)

X_train, X_test, y_train, y_test = train_test_split(X_transform_em, y, random_state=50, test_size=0.2)

In [ ]:
naive, random, decision_tree, svm, r_logistic, knn = inicializar_modelos()

In [ ]:
naive.fit(X_train, y_train)
random.fit(X_train, y_train)
decision_tree.fit(X_train, y_train)
svm.fit(X_train, y_train)
r_logistic.fit(X_train, y_train)
knn.fit(X_train, y_train)

KNeighborsClassifier(n_neighbors=3)

In [ ]:
random_result = random.predict(X_test)
naive_result = naive.predict(X_test)
decision_result = decision_tree.predict(X_test)
svm_result = svm.predict(X_test)
logistic_result = r_logistic.predict(X_test)
knn_result = knn.predict(X_test)

print("Random forest: " + acuracia(random_result, y_test))
print("Naive Bayes: " + acuracia(naive_result, y_test))
print("Decision Tree: " + acuracia(decision_result, y_test))
print("SVM: " + acuracia(svm_result, y_test))
print("Logistic Regression: "+acuracia(logistic_result, y_test))
print("KNN: "+acuracia(knn_result, y_test))

Random forest: Acurácia do modelo: 66.67 %
Naive Bayes: Acurácia do modelo: 70.83 %
Decision Tree: Acurácia do modelo: 58.33 %
SVM: Acurácia do modelo: 45.83 %
Logistic Regression: Acurácia do modelo: 54.17 %
KNN: Acurácia do modelo: 62.5 %


**-Segunda Análise: Aplicando Seleção de Atributos (Wrapper)**

In [ ]:
X_wrapper = wrapper_based(X, X_transform, y, 100)

X_wrapper

TX_RESP_Q030_B  TX_RESP_Q026_C  ...  IN_COMUM_EJA_MEDIO  QtdeDocentesFCEM
27000095               1               9  ...                   1                 0
27001237               4              19  ...                   1                31
27002020               5               9  ...                   1                 0
27002063               1               9  ...                   0                 0
27002950               4              14  ...                   1                 0
...                  ...             ...  ...                 ...               ...
27226131               4              30  ...                   1                 5
27226239               2              26  ...                   1                 0
27226468               4              13  ...                   1                 0
27226808               1               7  ...                   1                 6
27353036               4              17  ...                   0                59

[118 rows x 100 columns]

In [ ]:
X_transform_w = minmax(X_wrapper)

X_train, X_test, y_train, y_test = train_test_split(X_transform_w, y, random_state=50, test_size=0.2)

In [ ]:
naive, random, decision_tree, svm, r_logistic, knn = inicializar_modelos()

In [ ]:
naive.fit(X_train, y_train)
random.fit(X_train, y_train)
decision_tree.fit(X_train, y_train)
svm.fit(X_train, y_train)
r_logistic.fit(X_train, y_train)
knn.fit(X_train, y_train)

KNeighborsClassifier(n_neighbors=3)

In [ ]:
random_result = random.predict(X_test)
naive_result = naive.predict(X_test)
decision_result = decision_tree.predict(X_test)
svm_result = svm.predict(X_test)
logistic_result = r_logistic.predict(X_test)
knn_result = knn.predict(X_test)

print("Random forest: " + acuracia(random_result, y_test))
print("Naive Bayes: " + acuracia(naive_result, y_test))
print("Decision Tree: " + acuracia(decision_result, y_test))
print("SVM: " + acuracia(svm_result, y_test))
print("Logistic Regression: "+acuracia(logistic_result, y_test))
print("KNN: "+acuracia(knn_result, y_test))

Random forest: Acurácia do modelo: 62.5 %
Naive Bayes: Acurácia do modelo: 66.67 %
Decision Tree: Acurácia do modelo: 50.0 %
SVM: Acurácia do modelo: 75.0 %
Logistic Regression: Acurácia do modelo: 62.5 %
KNN: Acurácia do modelo: 66.67 %


**-Segunda Análise: Aplicando Seleção de Atributos (Filtragem)**

In [ ]:
X_selection = selecao_atributos(X, y, 100)

X_selection

/usr/local/lib/python3.7/dist-packages/sklearn/feature_selection/_univariate_selection.py:112: UserWarning: Features [  0   1   2 234 236 239 242 248 250 251 252 253 254 255 256 266 267 268
 269 272 275 276 277 278 291 292 306 310 325 349 356 357 359 360 363 370
 371 372 373 374 375 376 379 380 381 382 384 390 391] are constant.
  warnings.warn("Features %s are constant." % constant_features_idx, UserWarning)
/usr/local/lib/python3.7/dist-packages/sklearn/feature_selection/_univariate_selection.py:113: RuntimeWarning: invalid value encountered in true_divide
  f = msb / msw


ID_DEPENDENCIA_ADM  ...  QtdeDocentesTempo
27000095                   2  ...                148
27001237                   2  ...                225
27002020                   2  ...                217
27002063                   2  ...                 40
27002950                   2  ...                 84
...                      ...  ...                ...
27226131                   2  ...                195
27226239                   2  ...                363
27226468                   2  ...                190
27226808                   2  ...                 89
27353036                   2  ...                251

[118 rows x 100 columns]

In [ ]:
X_transform = minmax(X_selection)

X_train, X_test, y_train, y_test = train_test_split(X_transform, y, random_state=50, test_size=0.2)

**Fase de Treinamento e Teste**

In [ ]:
naive, random, decision_tree, svm, r_logistic, knn = inicializar_modelos()

In [ ]:
naive.fit(X_train, y_train)
random.fit(X_train, y_train)
decision_tree.fit(X_train, y_train)
svm.fit(X_train, y_train)
r_logistic.fit(X_train, y_train)
knn.fit(X_train, y_train)

KNeighborsClassifier(n_neighbors=3)

In [ ]:
random_result = random.predict(X_transform)
naive_result = naive.predict(X_transform)
decision_result = decision_tree.predict(X_transform)
svm_result = svm.predict(X_transform)
logistic_result = r_logistic.predict(X_transform)
knn_result = knn.predict(X_transform)

print("Random forest: " + acuracia(random_result, y))
print("Naive Bayes: " + acuracia(naive_result, y))
print("Decision Tree: " + acuracia(decision_result, y))
print("SVM: " + acuracia(svm_result, y))
print("Logistic Regression: "+acuracia(logistic_result, y))
print("KNN: "+acuracia(knn_result, y))

Random forest: Acurácia do modelo: 90.68 %
Naive Bayes: Acurácia do modelo: 61.02 %
Decision Tree: Acurácia do modelo: 88.14 %
SVM: Acurácia do modelo: 82.2 %
Logistic Regression: Acurácia do modelo: 80.51 %
KNN: Acurácia do modelo: 69.49 %


**Segunda Análise: Unindo subconjunto de melhores atributos**

In [ ]:
X_all = pd.concat([X_selection, X_wrapper], axis=1)

X_all = pd.concat([X_all, X_embedded], axis=1)

In [ ]:
#Removendo colunas duplicadas
X_new_all = X_all.T.drop_duplicates().T

In [ ]:
X_transform = minmax(X_new_all)
#X_transform = minmax(X_all)

X_train, X_test, y_train, y_test = train_test_split(X_transform, y, random_state=50, test_size=0.2)

In [ ]:
naive, random, decision_tree, svm, r_logistic, knn = inicializar_modelos()

In [ ]:
naive.fit(X_train, y_train)
random.fit(X_train, y_train)
decision_tree.fit(X_train, y_train)
svm.fit(X_train, y_train)
r_logistic.fit(X_train, y_train)
knn.fit(X_train, y_train)

KNeighborsClassifier(n_neighbors=3)

In [ ]:
random_result = random.predict(X_transform)
naive_result = naive.predict(X_transform)
decision_result = decision_tree.predict(X_transform)
svm_result = svm.predict(X_transform)
logistic_result = r_logistic.predict(X_transform)
knn_result = knn.predict(X_transform)

print("Random forest: " + acuracia(random_result, y))
print("Naive Bayes: " + acuracia(naive_result, y))
print("Decision Tree: " + acuracia(decision_result, y))
print("SVM: " + acuracia(svm_result, y))
print("Logistic Regression: "+acuracia(logistic_result, y))
print("KNN: "+acuracia(knn_result, y))

Random forest: Acurácia do modelo: 91.53 %
Naive Bayes: Acurácia do modelo: 64.41 %
Decision Tree: Acurácia do modelo: 90.68 %
SVM: Acurácia do modelo: 89.83 %
Logistic Regression: Acurácia do modelo: 90.68 %
KNN: Acurácia do modelo: 66.1 %


Resultado com Todos atributos

- Random forest: Acurácia do modelo: 92.37 %
- Naive Bayes: Acurácia do modelo: 66.1 %
- Decision Tree: Acurácia do modelo: 92.37 %
- SVM: Acurácia do modelo: 89.83 %
- Logistic Regression: Acurácia do modelo: 91.53 %
- KNN: Acurácia do modelo: 66.95 %

**Segunda Análise: Aplicando a técnica merge**

Link da Dissertação: https://www.dcc.ufmg.br/pos/cursos/defesas/1930M.PDF


In [ ]:
#Metodo Merge combina os atributos mais frequentes nos melhores conjuntos de seleção 
#Cada atributo irá apresentar um score de acordo com sua presença em um subconjunto de dados
def metodo_merge(selection, embedded, wrapper, ranking):
  for atributo in selection:
    score = ranking[atributo]
    if (atributo in embedded):
      score += 1
      embedded.remove(atributo)
    if (atributo in wrapper):
      score +=1
      wrapper.remove(atributo)
    ranking[atributo] = score

  for atributo in embedded:
    score = ranking[atributo]
    if (atributo in selection):
      score += 1
      embedded.remove(atributo)
    if (atributo in wrapper):
      score +=1
      wrapper.remove(atributo)
    ranking[atributo] = score

  for atributo in wrapper:
    score = ranking[atributo]
    if (atributo in selection):
      score += 1
      embedded.remove(atributo)
    if (atributo in embedded):
      score +=1
      wrapper.remove(atributo)
    ranking[atributo] = score
 
  return ranking      

In [ ]:
#Gerando Listas com os melhores atributos de cada abordagem
selection = list(X_selection.columns)
embedded = list(X_embedded.columns)
wrapper = list(X_wrapper.columns)

#Criando dicionário
values = [1]*300

all = selection + embedded + wrapper

ranking = dict(zip(all, values))

ranking = metodo_merge(selection, embedded, wrapper, ranking)

#Filtrando os atributos que estão em mais de um subconjunto

dic_aux = ranking.items()

filtered = {key:value for key, value in dic_aux if value > 1}

#Gerando X após o merge

X_merge = dataset[list(filtered.keys())]

In [ ]:
X_transform = minmax(X_merge)

X_train, X_test, y_train, y_test = train_test_split(X_transform, y, random_state=50, test_size=0.2)

In [ ]:
naive, random, decision_tree, svm, r_logistic, knn = inicializar_modelos()

In [ ]:
naive.fit(X_train, y_train)
random.fit(X_train, y_train)
decision_tree.fit(X_train, y_train)
svm.fit(X_train, y_train)
r_logistic.fit(X_train, y_train)
knn.fit(X_train, y_train)

KNeighborsClassifier(n_neighbors=3)

In [ ]:
random_result = random.predict(X_transform)
naive_result = naive.predict(X_transform)
decision_result = decision_tree.predict(X_transform)
svm_result = svm.predict(X_transform)
logistic_result = r_logistic.predict(X_transform)
knn_result = knn.predict(X_transform)

print("Random forest: " + acuracia(random_result, y))
print("Naive Bayes: " + acuracia(naive_result, y))
print("Decision Tree: " + acuracia(decision_result, y))
print("SVM: " + acuracia(svm_result, y))
print("Logistic Regression: "+acuracia(logistic_result, y))
print("KNN: "+acuracia(knn_result, y))

Random forest: Acurácia do modelo: 92.37 %
Naive Bayes: Acurácia do modelo: 68.64 %
Decision Tree: Acurácia do modelo: 90.68 %
SVM: Acurácia do modelo: 91.53 %
Logistic Regression: Acurácia do modelo: 91.53 %
KNN: Acurácia do modelo: 70.34 %


In [ ]:
dados_para_exportar = pd.DataFrame(columns=X_merge.columns, data=X_merge.values, index=X_merge.index)
dados_para_exportar.insert(len(dados_para_exportar.columns), "Valor Real Ideb", y.to_numpy())
dados_para_exportar.insert(len(dados_para_exportar.columns), "Previsão Ideb", random_result)


dados_para_exportar

TX_RESP_Q030_B  TX_RESP_Q026_C  ...  Valor Real Ideb    Previsão Ideb
27000095               1               9  ...            média            média
27001237               4              19  ...   acima da média   acima da média
27002020               5               9  ...  abaixo da média  abaixo da média
27002063               1               9  ...   acima da média   acima da média
27002950               4              14  ...   acima da média   acima da média
...                  ...             ...  ...              ...              ...
27226131               4              30  ...  abaixo da média  abaixo da média
27226239               2              26  ...  abaixo da média  abaixo da média
27226468               4              13  ...   acima da média   acima da média
27226808               1               7  ...   acima da média   acima da média
27353036               4              17  ...  abaixo da média  abaixo da média

[118 rows x 88 columns]

In [ ]:
dados_para_exportar = dados_para_exportar.loc[dados_para_exportar['Valor Real Ideb'] ==  dados_para_exportar['Previsão Ideb']]


dados_para_exportar

TX_RESP_Q030_B  TX_RESP_Q026_C  ...  Valor Real Ideb    Previsão Ideb
27000095               1               9  ...            média            média
27001237               4              19  ...   acima da média   acima da média
27002020               5               9  ...  abaixo da média  abaixo da média
27002063               1               9  ...   acima da média   acima da média
27002950               4              14  ...   acima da média   acima da média
...                  ...             ...  ...              ...              ...
27226131               4              30  ...  abaixo da média  abaixo da média
27226239               2              26  ...  abaixo da média  abaixo da média
27226468               4              13  ...   acima da média   acima da média
27226808               1               7  ...   acima da média   acima da média
27353036               4              17  ...  abaixo da média  abaixo da média

[109 rows x 88 columns]

In [ ]:
#dados_para_exportar.to_csv('C:/Users/maril/Documents/ufs/TCC e Pesquisa IDEB/resultado_analise_merge_dataset_completo.csv')
#dados_para_exportar.to_csv('/content/drive/MyDrive/TCC/resultado_analise_merge_dataset_completo.csv')
dados_para_exportar.to_csv('/content/drive/MyDrive/TCC/resultado_analise_merge_dataset_completo_AL.csv')

#### -Terceira análise: variáveis das escolas e professores

In [ ]:
columns_escolas = dataset.columns[237:404]

X_escolas = dataset[columns_escolas]

X_escolas

TP_SITUACAO_FUNCIONAMENTO  ...  Ideb-2017
27000095                          1  ...        3.4
27001237                          1  ...        3.9
27002020                          1  ...        3.3
27002063                          1  ...        3.8
27002950                          1  ...        3.5
...                             ...  ...        ...
27226131                          1  ...        2.8
27226239                          1  ...        3.1
27226468                          1  ...        4.1
27226808                          1  ...        3.9
27353036                          1  ...        3.0

[118 rows x 164 columns]

In [ ]:
X_escolas.drop(columns=['DT_ANO_LETIVO_INICIO', 'DT_ANO_LETIVO_TERMINO'], axis=1, inplace=True)

X_transform = minmax(X_escolas)

X_train, X_test, y_train, y_test = train_test_split(X_transform, y, random_state=50, test_size=0.2)

/usr/local/lib/python3.7/dist-packages/pandas/core/frame.py:4913: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  errors=errors,


In [ ]:
naive, random, decision_tree, svm, r_logistic, knn = inicializar_modelos()

In [ ]:
naive.fit(X_train, y_train)
random.fit(X_train, y_train)
decision_tree.fit(X_train, y_train)
svm.fit(X_train, y_train)
r_logistic.fit(X_train, y_train)
knn.fit(X_train, y_train)

KNeighborsClassifier(n_neighbors=3)

In [ ]:
random_result = random.predict(X_test)
naive_result = naive.predict(X_test)
decision_result = decision_tree.predict(X_test)
svm_result = svm.predict(X_test)
logistic_result = r_logistic.predict(X_test)
knn_result = knn.predict(X_test)

print("Random forest: " + acuracia(random_result, y_test))
print("Naive Bayes: " + acuracia(naive_result, y_test))
print("Decision Tree: " + acuracia(decision_result, y_test))
print("SVM: " + acuracia(svm_result, y_test))
print("Logistic Regression: "+acuracia(logistic_result, y_test))
print("KNN: "+acuracia(knn_result, y_test))

Random forest: Acurácia do modelo: 83.33 %
Naive Bayes: Acurácia do modelo: 50.0 %
Decision Tree: Acurácia do modelo: 100.0 %
SVM: Acurácia do modelo: 37.5 %
Logistic Regression: Acurácia do modelo: 45.83 %
KNN: Acurácia do modelo: 45.83 %


**Aplicando seleção de atributos (Filtragem)**

In [ ]:
X_selection = selecao_atributos(X_escolas, y, 100)

/usr/local/lib/python3.7/dist-packages/sklearn/feature_selection/_univariate_selection.py:112: UserWarning: Features [  0   1   4  10  12  13  14  15  16  17  18  28  29  30  31  34  37  38
  39  40  53  54  68  72  87 111 118 119 121 122 125 132 133 134 135 136
 137 138 141 142 143 144 146 152 153] are constant.
  warnings.warn("Features %s are constant." % constant_features_idx, UserWarning)
/usr/local/lib/python3.7/dist-packages/sklearn/feature_selection/_univariate_selection.py:113: RuntimeWarning: invalid value encountered in true_divide
  f = msb / msw


In [ ]:
naive, random, decision_tree, svm, r_logistic, knn = inicializar_modelos()

In [ ]:
naive.fit(X_train, y_train)
random.fit(X_train, y_train)
decision_tree.fit(X_train, y_train)
svm.fit(X_train, y_train)
r_logistic.fit(X_train, y_train)
knn.fit(X_train, y_train)

KNeighborsClassifier(n_neighbors=3)

In [ ]:
random_result = random.predict(X_test)
naive_result = naive.predict(X_test)
decision_result = decision_tree.predict(X_test)
svm_result = svm.predict(X_test)
logistic_result = r_logistic.predict(X_test)
knn_result = knn.predict(X_test)

print("Random forest: " + acuracia(random_result, y_test))
print("Naive Bayes: " + acuracia(naive_result, y_test))
print("Decision Tree: " + acuracia(decision_result, y_test))
print("SVM: " + acuracia(svm_result, y_test))
print("Logistic Regression: "+acuracia(logistic_result, y_test))
print("KNN: "+acuracia(knn_result, y_test))

Random forest: Acurácia do modelo: 79.17 %
Naive Bayes: Acurácia do modelo: 50.0 %
Decision Tree: Acurácia do modelo: 100.0 %
SVM: Acurácia do modelo: 37.5 %
Logistic Regression: Acurácia do modelo: 45.83 %
KNN: Acurácia do modelo: 45.83 %


**Aplicando seleção de atributos (Embutida)**

In [ ]:
X_embedded = embedded_based(X, X_transform, y, 100)

X_embedded

/usr/local/lib/python3.7/dist-packages/sklearn/linear_model/_logistic.py:818: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG,


ID_UF  ID_MUNICIPIO  ...  TX_RESP_Q013_A  TX_RESP_Q013_C
27000095     27       2700102  ...              80               1
27001237     27       2701605  ...             145               6
27002020     27       2705002  ...              78               2
27002063     27       2705002  ...              30               0
27002950     27       2706422  ...             112               1
...         ...           ...  ...             ...             ...
27226131     27       2700706  ...             107               2
27226239     27       2702306  ...              90               1
27226468     27       2708303  ...             106               9
27226808     27       2707602  ...              65               1
27353036     27       2709152  ...             130               1

[118 rows x 65 columns]

In [ ]:
X_transform_em = minmax(X_embedded)

X_train, X_test, y_train, y_test = train_test_split(X_transform_em, y, random_state=50, test_size=0.2)

In [ ]:
naive, random, decision_tree, svm, r_logistic, knn = inicializar_modelos()

In [ ]:
naive.fit(X_train, y_train)
random.fit(X_train, y_train)
decision_tree.fit(X_train, y_train)
svm.fit(X_train, y_train)
r_logistic.fit(X_train, y_train)
knn.fit(X_train, y_train)

KNeighborsClassifier(n_neighbors=3)

In [ ]:
random_result = random.predict(X_test)
naive_result = naive.predict(X_test)
decision_result = decision_tree.predict(X_test)
svm_result = svm.predict(X_test)
logistic_result = r_logistic.predict(X_test)
knn_result = knn.predict(X_test)

print("Random forest: " + acuracia(random_result, y_test))
print("Naive Bayes: " + acuracia(naive_result, y_test))
print("Decision Tree: " + acuracia(decision_result, y_test))
print("SVM: " + acuracia(svm_result, y_test))
print("Logistic Regression: "+acuracia(logistic_result, y_test))
print("KNN: "+acuracia(knn_result, y_test))

Random forest: Acurácia do modelo: 50.0 %
Naive Bayes: Acurácia do modelo: 62.5 %
Decision Tree: Acurácia do modelo: 41.67 %
SVM: Acurácia do modelo: 58.33 %
Logistic Regression: Acurácia do modelo: 70.83 %
KNN: Acurácia do modelo: 41.67 %


**Aplicando seleção de atributos (Wrapper)**

In [ ]:
X_wrapper = wrapper_based(X, X_transform, y, 100)

X_wrapper

ID_UF  ID_MUNICIPIO  ...  TX_RESP_Q013_B  TX_RESP_Q013_C
27000095     27       2700102  ...              15               1
27001237     27       2701605  ...              34               6
27002020     27       2705002  ...              20               2
27002063     27       2705002  ...               6               0
27002950     27       2706422  ...              29               1
...         ...           ...  ...             ...             ...
27226131     27       2700706  ...              45               2
27226239     27       2702306  ...              27               1
27226468     27       2708303  ...              37               9
27226808     27       2707602  ...              12               1
27353036     27       2709152  ...              45               1

[118 rows x 100 columns]

In [ ]:
X_transform_w = minmax(X_wrapper)

X_train, X_test, y_train, y_test = train_test_split(X_transform_w, y, random_state=50, test_size=0.2)

In [ ]:
naive, random, decision_tree, svm, r_logistic, knn = inicializar_modelos()

In [ ]:
naive.fit(X_train, y_train)
random.fit(X_train, y_train)
decision_tree.fit(X_train, y_train)
svm.fit(X_train, y_train)
r_logistic.fit(X_train, y_train)
knn.fit(X_train, y_train)

KNeighborsClassifier(n_neighbors=3)

In [ ]:
random_result = random.predict(X_test)
naive_result = naive.predict(X_test)
decision_result = decision_tree.predict(X_test)
svm_result = svm.predict(X_test)
logistic_result = r_logistic.predict(X_test)
knn_result = knn.predict(X_test)

print("Random forest: " + acuracia(random_result, y_test))
print("Naive Bayes: " + acuracia(naive_result, y_test))
print("Decision Tree: " + acuracia(decision_result, y_test))
print("SVM: " + acuracia(svm_result, y_test))
print("Logistic Regression: "+acuracia(logistic_result, y_test))
print("KNN: "+acuracia(knn_result, y_test))

Random forest: Acurácia do modelo: 54.17 %
Naive Bayes: Acurácia do modelo: 58.33 %
Decision Tree: Acurácia do modelo: 50.0 %
SVM: Acurácia do modelo: 54.17 %
Logistic Regression: Acurácia do modelo: 54.17 %
KNN: Acurácia do modelo: 41.67 %
